src/modules/Game.class.js

```diff
@@ -20,25 +20,229 @@ class Game {
    * If passed, the board will be initialized with the provided
    * initial state.
    */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
+  constructor(initialState = null) {
+    this.size = 4;
+
+    this.initialState = initialState
+      ? initialState.map((row) => [...row])
+      : null;
+
+    this.board = initialState
+      ? initialState.map((row) => [...row])
+      : this.createEmptyBoard();
+    this.score = 0;
+    this.status = 'idle';
+  }
+
+  createEmptyBoard() {
+    return Array.from({ length: this.size }, () => Array(this.size).fill(0));
+  }
+
+  addRandomTile() {
+    const emptyCells = [];
+
+    for (let r = 0; r < this.size; r++) {
+      for (let c = 0; c < this.size; c++) {
+        if (this.board[r][c] === 0) {
+          emptyCells.push([r, c]);
+        }
+      }
+    }
+
+    if (emptyCells.length === 0) {
+      return;
+    }
+
+    const [row, column] =
+      emptyCells[Math.floor(Math.random() * emptyCells.length)];
+
+    this.board[row][column] = Math.random() < 0.9 ? 2 : 4;
+  }
+
+  moveLeft() {
+    if (this.status !== 'playing') {
+      return;
+    }
+
+    let moved = false;
+
+    for (let r = 0; r < this.size; r++) {
+      const currentRow = this.board[r];
+      const newRow = this.combineRow(currentRow);
+
+      for (let i = 0; i < this.size; i++) {
+        if (currentRow[i] !== newRow[i]) {
+          moved = true;
+          break;
+        }
+      }
+
+      this.board[r] = newRow;
+    }
+
+    if (moved) {
+      this.afterMove();
+    }
```

Comment:
This logic is similar in every method, try to put it in a helper function and use it everywhere.